### PR TITLE
Update bencher code for Linq To DB

### PR DESCRIPTION
**Pull request by MaceWindu**
- enable all benchmarks except unsupported functionality like change tracking
- add missing mappings
- cleanup code

### CODE REVIEW

**Comment by FransBouma** [comment]
Your description is way too small for the amount of changes made. So please explain why these changes were all needed. E.g. 'missing' mappings, it's a microORM so I don't assume these missing mappings were needed?

**Comment by MaceWindu** [comment]
Existing linq2db bencher implements fetch-only tests. I've updated it to support other benchmarks (eager load, async, inserts) and defined mappings for tables, used in those tests. They were missing, because they are not used for fetch tests.

File: RawBencher/Benchers/LINQ2DBNormalBencher.cs [hunk 1]
```diff
@@ -1,47 +1,153 @@
...
+                    .Where(_ => _.CreditCardID > 19237)
+                   .ToList();
+            }
+        }
+
+        protected override void DeleteInserted(IEnumerable<CreditCard> toDelete)
+        {
+            using (var db = new Db(ConnectionString))
+            {
+                // other providers cheat by using "CreditCardId > 19237" filter
```
Comment (FransBouma) [inline RawBencher/Benchers/LINQ2DBNormalBencher.cs]: It's not used in measurements, so you can use whatever you want to get the rows as quickly as possible from the DB. So if a single delete with a single predicate does the trick, that's fine, the faster the better as e.g. with EF it really slows things down as they have to fetch each entity that's to be deleted.

**Comment by FransBouma** [comment]
```
Set inserts of 1000 elements in one go (10 runs with batchsize 100)
------------------------------------------------------------------------------
LINQ to DB v2.7.0.0 (v2.7.0) (normal)                                 : 30,67ms (2,92ms)
LINQ to DB v2.7.0.0 (v2.7.0) (compiled)                               : 37,41ms (5,01ms)
LLBLGen Pro v5.5.0.0 (v5.5.3)                                         : 64,21ms (1,00ms)
Entity Framework Core v2.2.4.0 (v2.2.4.19081)                         : 97,66ms (2,20ms)

Memory usage, per iteration
------------------------------------------------------------------------------
LINQ to DB v2.7.0.0 (v2.7.0) (normal)                                 : 328 KB (336.304 bytes)
LINQ to DB v2.7.0.0 (v2.7.0) (compiled)                               : 328 KB (336.304 bytes)
LLBLGen Pro v5.5.0.0 (v5.5.3)                                         : 11.915 KB (12.201.928 bytes)
Entity Framework Core v2.2.4.0 (v2.2.4.19081)                         : 16.790 KB (17.193.792 bytes)
```
Two things:
1) the batch inserts aren't batch inserts, but something else. I have to check what's sent to the db but I expect sqlbulkcopy statements. The batching benchmark isn't meant for that, it's meant to show batching of SQL statements, be it inserts or updates, to e.g. different tables, not the time it takes to bypass the SQL engine and do a bulk copy. I know inserting 1000 rows using the SQL engine/pipeline might sound like a job that's better suited for sqlbulkcopy but the bencher isn't meant to show sql server's features but what the orm can do, so if it offers batching, it should rely on the sql engine/pipeline and not on a bulkcopy feature offered by the db.
2) the memory usage of the batching in linq2db is really low compared to others. 1000 entities take up more memory than that, so why is the amount of memory used in your framework so low compared to the others? I don't expect you to consume 11MB, as my framework boxes a bit, so I know why it's 11,9MB. I don't know why EF needs that much memory, so let's leave that aside. But 300KB is a bit low.

**Comment by FransBouma** [comment]
```
insert bulk [Sales].[CreditCard] ([CardType] NVarChar(50) COLLATE SQL_Latin1_General_CP1_CI_AS, [CardNumber] NVarChar(25) COLLATE SQL_Latin1_General_CP1_CI_AS, [ExpMonth] TinyInt, [ExpYear] SmallInt, [ModifiedDate] DateTime)
```
Yes, this isn't going to work in the batch insert test, I'm afraid. It's a nice feature and I congratulate you for implementing it :) (I recon using SqlClient's SqlBulkCopy.WriteToServer? ) But it's not the point of the batch insert benchmark. So before the PR can be merged, I'm afraid the batch insert benchmark you have added has to be removed.

**Comment by MaceWindu** [comment]
Right, it was using bulk copy. Updated code to use inserts

**Comment by FransBouma** [comment]
The inserts give an exception: 

```
Exception caught of type: System.Data.SqlClient.SqlException
Message: Cannot insert explicit value for identity column in table 'CreditCard' when IDENTITY_INSERT is set to OFF.
Stack trace:
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite, String methodName)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at LinqToDB.Data.DataConnection.ExecuteNonQuery() in C:\projects\linq2db\Source\LinqToDB\Data\DataConnection.cs:line 1136
   at LinqToDB.Data.CommandInfo.Execute() in C:\projects\linq2db\Source\LinqToDB\Data\CommandInfo.cs:line 471
   at LinqToDB.DataProvider.MultipleRowsHelper.Execute() in C:\projects\linq2db\Source\LinqToDB\DataProvider\MultipleRowsHelper.cs:line 98
   at LinqToDB.DataProvider.BasicBulkCopy.MultipleRowsCopy1(MultipleRowsHelper helper, IEnumerable source) in C:\projects\linq2db\Source\LinqToDB\DataProvider\BasicBulkCopy.cs:line 259
   at LinqToDB.DataProvider.SqlServer.SqlServerBulkCopy.MultipleRowsCopy[T](ITable`1 table, BulkCopyOptions options, IEnumerable`1 source) in C:\projects\linq2db\Source\LinqToDB\DataProvider\SqlServer\SqlServerBulkCopy.cs:line 112
   at RawBencher.Benchers.LINQ2DBNormalBencher.InsertSet(IEnumerable`1 toInsert, Int32 batchSize) in C:\MyProjects\VS.NET Projects\GitHub\FransBouma\RawDataAccessBencher\RawBencher\Benchers\LINQ2DBNormalBencher.cs:line 146
   at RawBencher.Benchers.BencherBase`2.PerformInsertSetBenchmark(Int32 amountToInsert, Int32 batchSize, Boolean discardResults) in C:\MyProjects\VS.NET Projects\GitHub\FransBouma\RawDataAccessBencher\RawBencher\Benchers\BencherBase.cs:line 318
   at RawBencher.BenchController.RunMemoryAnalysisForBencher(IBencher bencher) in C:\MyProjects\VS.NET Projects\GitHub\FransBouma\RawDataAccessBencher\RawBencher\BenchController.cs:line 359
   at RawBencher.BenchController.RunRegisteredBenchers() in C:\MyProjects\VS.NET Projects\GitHub\FransBouma\RawDataAccessBencher\RawBencher\BenchController.cs:line 279
Inner exception:
```
easy fix I think on your end :)

**Comment by MaceWindu** [comment]
my bad, should work now

**Comment by FransBouma** [comment]
All good. I see you pass a multi-value list, neat. I'll allow it. Yours uses that, mine uses concatted insert statements, EF Core uses a giant MERGE statement. Of course yours isn't as flexible (inserting in 10 tables in a single batch will likely paint a picture of slower performance but still... ), but it's still a neat way to do batching. I'll merge the PR (locally as I've made a fix to the reporting code where it missed a `\n`)

Thanks for the contribution and nice job :)

**Comment by MaceWindu** [comment]
Yeah, there are multiple ways to insert data, wasn't sure which one to choose for this test

**Pull request merged**